src/core/server/application/application_context_provider.ts

```diff
@@ -0,0 +1,181 @@
+
+  constructor(params: Params) {
+    Object.assign(this, params);
+  }
+
+  private async getUiPlugins(): Promise<ConfiguredPlugin[]> {
+    return Promise.all(
+      [...this.plugins.uiPlugins.public].map(async ([id, plugin]) => {
+        const config = await this.plugins.uiPluginConfigs
+          .get(id)
+          ?.pipe(take(1))
+          .toPromise();
+
+        return { id, plugin, config: config || {} };
+      })
+    );
+  }
+
+  private async getMetadata(
+    injectedOverrides: PluginVariables,
+    includeUserProvidedConfig: boolean
+  ) {
+    const [uiPlugins, uiSettings, userSettings] = await Promise.all<
+      ConfiguredPlugin[],
+      PluginVariables,
+      UserVariables
+    >([
+      this.getUiPlugins(),
+      this.uiSettings.getAll(),
+      includeUserProvidedConfig ? this.uiSettings.getUserProvided() : Promise.resolve({}),
+    ]);
+
+    return {
+      uiPlugins,
+      uiSettings,
+      userSettings,
+      vars: mergeVariables(
+        uiPlugins.map(({ config }) => config),
+        injectedOverrides
+      ),
+    };
+  }
+
+  private async getVariables(
+    pluginId: string,
+    injectedOverrides: PluginVariables,
+    includeUserProvidedConfig: boolean
+  ) {
+    const { uiPlugins, uiSettings, userSettings, vars } = await this.getMetadata(
+      injectedOverrides,
+      includeUserProvidedConfig
+    );
+    const basePath = this.http.basePath.get(this.request);
+    // TODO: Use full implementation once https://github.com/elastic/kibana/pull/51438 lands:
+    const capabilities = defaultCapabilities;
+
+    return {
+      uiPublicUrl: `${basePath}/ui`,
+      bootstrapScriptUrl: `${basePath}/bundles/app/${pluginId}/bootstrap.js`,
+      i18n: i18n.translate,
+      locale: i18n.getLocale(),
+      darkMode: userSettings['theme:darkMode']?.userValue ?? false,
+      injectedMetadata: {
+        version: this.http.server.version,
+        buildNumber: this.env.packageInfo.buildNum,
+        branch: this.env.packageInfo.branch,
+        basePath,
+        env: this.env,
+        legacyMode: false,
+        i18n: {
+          translationsUrl: `${basePath}/translations/${i18n.getLocale()}.json`,
+        },
+        csp: this.http.csp,
+        vars,
+        uiPlugins,
+        capabilities,
+        legacyMetadata: {
+          app: {},
+          bundleId: `app:${pluginId}`,
+          nav: capabilities.navLinks,
+          env: this.env,
+          version: this.env.packageInfo.version,
+          branch: this.env.packageInfo.branch,
+          buildNum: this.env.packageInfo.buildNum,
+          buildSha: this.env.packageInfo.buildSha,
+          serverName: this.http.server.name,
+          devMode: this.env.mode.dev,
+          basePath,
+          uiSettings: {
+            defaults: uiSettings,
+            user: userSettings,
+          },
+        },
+      },
+    };
+  }
+
+  public async render(id = 'core', injectedVars = {}, includeUserProvidedConfig = true) {
```

Comment:
Just reiterating here so it doesn't get lost. New platform plugins should always render the same bundle, so they do not need to be able to specify an `id`.

Would it be possible to re-use this logic and call it in the legacy `ui_render_mixin` for rendering legacy apps too? I worry about the implementations diverging and causing subtle bugs that aren't caught by tests.

Reply:
I can certainly do that, but I want to temper that against our original decision to copy the logic for the sake of time and the fact that legacy is being removed. I'd also be happy to swap this out in a follow-up.

Reply:
We should decide on whether `render` supports LP or not. Anyway, we should remove `id` and  `injectedVars` from the **public** interface. 
Also +1 for switching to NP rendering mechanism in LP.

Reply:
I know I originally said we should just copy this, but after thinking about it some more I think it makes sense to spend the extra time here to use a single implementation for NP and LP because:
- The LP code is untested and any changes to it may cause unexpected breakages.
- This is a critical path to get right, which makes any test coverage that we can add very valuable.
- The legacy `ui_render_mixin` has been edited in 42 commits, which [tends to point to buggy code](https://google-engtools.blogspot.com/2011/12/bug-prediction-at-google.html), or at the very least, tightly coupled code. Tightly coupled code tends to be very hard to test.

Reply:
Sounds good, and just verified with @jinmu03 that we can move forward with the legacy bits in this PR.

Reply:
I'll commit my current changes and do the legacy switchover after that.

Reply:
https://github.com/elastic/kibana/pull/51527#issuecomment-561340383